### PR TITLE
Match CPython's islower/isupper exactly

### PR DESCRIPTION
**Pull request by joshuamegnauth54**
This PR fixes a regression from my last `islower`/`isupper` patch. Python's Bytes doesn't assume an encoding, so methods like `islower` should only consider ASCII casing.

I updated `islower`/`isupper` for UTF-8 and WTF-8 to match CPython exactly. The two functions now use the same properties as CPython and in the exact same way that CPython.

I updated the unit tests to pass on Python 3.15. Unicode updates sometimes cause properties to shift. I previously tested everything on Python 3.14, but that lead to failures that I assumed were bugs but were actually due to Unicode differences. For example, U+0295 is a lower case letter in older Unicode versions but is NOT in newer versions.

---

With Python 3.14:

```python
assert "\u0295".islower()
assert not "\u0295".isupper()
```

With Python 3.15 **and** RustPython:
```python
assert not "\u0295".islower()
assert not "\u0295".isupper()
assert not "\u0295".istitle()
```

^It's imperative to test with the latest Python because of Unicode differences. I, uh, spent like three hours trying to fix a bug that didn't exist because I was crosschecking the tests with Python 3.14... Oops. ~~I bumped CI's Python version to 3.15 as well.~~ Reverted because it's more complicated than I thought.

Here is [CPython's isupper implementation](https://github.com/python/cpython/blob/main/Objects/unicodeobject.c#L11774-L11805) for comparison.

CPython's core logic for `islower` (`isupper` is basically the same)
```c
/* Special case for empty strings */
if (length == 0)
    Py_RETURN_FALSE;

cased = 0;
for (i = 0; i < length; i++) {
    const Py_UCS4 ch = PyUnicode_READ(kind, data, i);

    if (Py_UNICODE_ISUPPER(ch) || Py_UNICODE_ISTITLE(ch))
        Py_RETURN_FALSE;
    else if (!cased && Py_UNICODE_ISLOWER(ch))
        cased = 1;
}
return PyBool_FromLong(cased);
```

RustPython's `islower`:

```rust
let mut all_cased = false;
for chunk in self.as_bytes().utf8_chunks().map(|c| c.valid()) {
    if chunk.chars().any(|c| {
        INVALID::for_char(c)
            || GeneralCategoryGroup::TitlecaseLetter.contains(GeneralCategory::for_char(c))
    }) {
        return false;
    }
    if !all_cased && chunk.chars().any(VALID::for_char) {
        all_cased = true;
    }
}
all_cased
```
They're the same now! :grin: 

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved case-checking: text strings use Unicode casedness detection while byte/encoded checks use ASCII-only scanning, preventing misclassification for mixed/invalid byte sequences and certain rare or non-cased characters.

* **Tests**
  * Updated and expanded tests for Unicode edge cases, version-dependent casing, non-cased scripts, and bytes case-checking with invalid/mixed sequences.

### CODE REVIEW

File: extra_tests/snippets/builtin_str_encode.py [hunk 1]
```diff
@@ -20,3 +20,13 @@ def round_trip(s, encoding="utf-8"):
 round_trip("👺♦  𝐚Şđƒ  ☆☝")
 round_trip("☢🐣  ᖇ𝓤𝕊тⓟ𝕐𝕥卄σ𝔫  ♬👣")
 round_trip("💀👌  ק𝔂tℍⓞ𝓷 ３  🔥👤")
+
+# Bytes should not assume an encoding for isupper/islower
+assert not "\u0001f431 CAT".encode().isupper()
+assert "\u0001f431 cat".encode().islower()
```
[the remark this answers is not in the document]
Comment (joshuamegnauth54) [reply extra_tests/snippets/builtin_str_encode.py]: I updated these tests to use better values and also fixed the mistake mentioned in the review.

File: crates/vm/src/stdlib/_thread.rs [hunk 2]
```diff
@@ -466,7 +466,7 @@ pub(crate) mod _thread {
         {
             // On Unix, use pthread ID from the handle
             use std::os::unix::thread::JoinHandleExt;
-            handle.as_pthread_t() as u64
+            handle.as_pthread_t()
```
[the remark this answers is not in the document]
Comment (joshuamegnauth54) [reply crates/vm/src/stdlib/_thread.rs]: Clippy complains about the explicit cast on Linux. Changing it to `as _` works for everything, I think.

**Comment by fanninpm** [comment]
Are there any relevant tests in, say, `Lib/test/string_tests.py` or `Lib/test/test_str.py`?

**Comment by joshuamegnauth54** [comment]
It looks like `extra_tests/snippets/stdlib_imp.py` fails with Python 3.15 but passes with Python 3.14. :thinking: I'll take a look at this later to see if I can fix it.

@fanninpm I skimmed them but couldn't find any tests for Unicode properties. It makes sense, though, because the properties may change across Unicode versions (e.g. \u0295 above). RustPython can *probably* remove its Unicode tests too once its string implementations match CPython exactly.

**Review by youknowone — APPROVED** [review]
Thanks!
